File: 6_Kyu_Kata/simple-substitution-cipher-helper/JavaScript/simpleSubstitutionCipherHelper.Test.js
// [object Object] - Simple Substitution Cipher Helper  [ ID: 52eb114b2d55f0e69800078d  (simple-substitution-cipher-helper) ]
// URL: https://www.codewars.com/kata/52eb114b2d55f0e69800078d
// Category: ALGORITHMS  |  Tags: CIPHERS | SECURITY | OBJECT-ORIENTED PROGRAMMING | STRINGS | ALGORITHMS
// *****************************************************************************
let abc1 = "abcdefghijklmnopqrstuvwxyz";
let abc2 = "etaoinshrdlucmfwypvbgkjqxz";
let sub = new SubstitutionCipher(abc1, abc2);
Test.assertEquals(sub.encode("abc"), "eta");
Test.assertEquals(sub.encode("xyz"), "qxz");
Test.assertEquals(sub.encode("aeiou"), "eirfg");
Test.assertEquals(sub.decode("eta"), "abc");
Test.assertEquals(sub.decode("qxz"), "xyz");
Test.assertEquals(sub.decode("eirfg"), "aeiou");

abc2 = 'tfozcivbsjhengarudlkpwqxmy';
sub = new SubstitutionCipher(abc1, abc2);
Test.assertEquals(sub.encode('abc'), 'tfo');
Test.assertEquals(sub.decode('tfo'), 'abc');
Test.assertEquals(sub.encode('tjuukf'), 'kjpphi');
Test.assertEquals(sub.decode('kjpphi'), 'tjuukf');
Test.assertEquals(sub.decode('tjuukf'), 'ajqqtb');
Test.assertEquals(sub.encode('ajqqtb'), 'tjuukf');

Test.assertEquals(sub.encode('a_bc&*83'), 't_fo&*83');
Test.assertEquals(sub.decode('t_fo*&83'), 'a_bc*&83');

abc1 = Test.randomize(abc1.split('')).join('');
abc2 = Test.randomize(abc1.split('')).join('');

sub = new SubstitutionCipher(abc1, abc2);
for (let i = 0; i < 8; i++) {
  let str = Test.randomToken();
  let cmp = str.split('').map(function (x) {
    let i = abc1.indexOf(x);
    return i < 0 ? x : abc2[i];
  }).join('');
  Test.assertEquals(sub.encode(str), cmp);
  cmp = str.split('').map(function (x) {
    let i = abc2.indexOf(x);
    return i < 0 ? x : abc1[i];
  }).join('');
  Test.assertEquals(sub.decode(str), cmp);
}

File: 6-kyu-kata/simple-substitution-cipher-helper/python/simple_substitution_cipher_helper_test.py
# 6 kyu - Simple Substitution Cipher Helper  [ ID: 52eb114b2d55f0e69800078d  (simple-substitution-cipher-helper) ]
# URL: https://www.codewars.com/kata/52eb114b2d55f0e69800078d
# Category: ALGORITHMS  |  Tags: CIPHERS | SECURITY | OBJECT-ORIENTED PROGRAMMING | STRINGS | ALGORITHMS
# ******************************************************************************
map1 = "abcdefghijklmnopqrstuvwxyz"
map2 = "etaoinshrdlucmfwypvbgkjqxz"

print("Examples")
cipher = Cipher(map1, map2)
test.assert_equals(cipher.encode("abc"), "eta")
test.assert_equals(cipher.encode("xyz"), "qxz")
test.assert_equals(cipher.encode("aeiou"), "eirfg")
test.assert_equals(cipher.decode("eta"), "abc")
test.assert_equals(cipher.decode("qxz"), "xyz")
test.assert_equals(cipher.decode("eirfg"), "aeiou")

map2 = "tfozcivbsjhengarudlkpwqxmy"
cipher = Cipher(map1, map2)
test.assert_equals(cipher.encode("abc"), "tfo")
test.assert_equals(cipher.decode("tfo"), "abc")
test.assert_equals(cipher.encode("tjuukf"), "kjpphi")
test.assert_equals(cipher.decode("kjpphi"), "tjuukf")
test.assert_equals(cipher.decode("tjuukf"), "ajqqtb")
test.assert_equals(cipher.encode("ajqqtb"), "tjuukf")

print("\nSpecial Characters")
test.assert_equals(cipher.encode("a_bc&*83"), "t_fo&*83")
test.assert_equals(cipher.decode("t_fo*&83"), "a_bc*&83")

print("\nRandomized tests")
import string, random


def string_shuffle(s):
    l = list(s)
    random.shuffle(l)
    return "".join(l)


map1 = string_shuffle(map1)
map2 = string_shuffle(map2)
encode_table = str.maketrans(map1, map2)
decode_table = str.maketrans(map2, map1)
cipher = Cipher(map1, map2)
for i in range(0, 10):
    s = "".join(random.sample(string.ascii_lowercase, 5))
    test.assert_equals(cipher.encode(s), s.translate(encode_table))

    s = "".join(random.sample(string.ascii_lowercase, 5))
    test.assert_equals(cipher.decode(s), s.translate(decode_table))
